File: androidApp/src/main/java/com/luisma/conexiones/android/features/game/GameViewState.kt
package com.luisma.conexiones.android.features.game

import com.luisma.conexiones.contracts.GAME_WORD_AMOUNT
import com.luisma.conexiones.models.BasicScreenState
import com.luisma.conexiones.models.game.GameData
import com.luisma.conexiones.models.game.GameDistributionCoordinates
import com.luisma.conexiones.models.game.GameGridRowState
import com.luisma.conexiones.models.game.GameState


data class GameViewState(
    val screenState: BasicScreenState,
    val gameData: GameData,
    val gameId: Int,
    val gridRowsState: Map<Int, GameGridRowState>,
    val currentSelection: List<GameDistributionCoordinates>,
    val lives: Int,
    val selectFailed: Boolean,
    val livesEarnedOnDone: Int,
    val nextGameId: Int,
    val notSolvedRowsOnLost: Map<Int, Int>,
    val showOnDoneSignAnimation: Boolean,
    val showOnDoneLivesAnimation: Boolean,
) {
    companion object {
        fun initial(): GameViewState {
            return GameViewState(
                screenState = BasicScreenState.Loading,
                gameData = GameData.initial(),
                gridRowsState = emptyMap(),
                currentSelection = emptyList(),
                gameId = -1,
                livesEarnedOnDone = 0,
                nextGameId = -1,
                lives = 0,
                selectFailed = false,
                notSolvedRowsOnLost = emptyMap(),
                showOnDoneSignAnimation = false,
                showOnDoneLivesAnimation = false
            )
        }
    }

    val canSelectWords: Boolean
        get() = gameData.gameState == GameState.Playing
                && currentSelection.count() != GAME_WORD_AMOUNT


    val canDeselectWords: Boolean
        get() = gameData.gameState == GameState.Playing
                && currentSelection.isNotEmpty()


    val canSendRow: Boolean
        get() = gameData.gameState == GameState.Playing
                && !selectFailed
                && currentSelection.count() == GAME_WORD_AMOUNT

    val showButtonsOnDone: Boolean
        get() {
            val state = gameData.gameState
            if ((state != GameState.Win && state != GameState.Lost) ||
                nextGameId == -1
            ) {
                return false
            }

            return if (state == GameState.Win) {
                !showOnDoneSignAnimation && !showOnDoneLivesAnimation
            } else {
                !showOnDoneSignAnimation
            }
        }

}

File: iosApp/iosApp/Features/Game/GameViewState.swift
//
//  GameViewState.swift
//  iosApp
//
//  Created by Luis Manuel García Fernández on 24/10/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import shared
import SwiftUI

struct GameViewState {
    var screenState: BasicScreenState
    var gameData: GameData
    var gameId: Int
    var gridRowsState: [Int: GameGridRowState]
    var currentSelection: [GameDistributionCoordinates]
    var lives: Int
    var selectFailed: Bool
    var livesEarnedOnDone: Int
    var nextGameId: Int
    var notSolvedRowsOnLost: [Int: Int]
    var showOnDoneSignAnimation: Bool
    var showOnDoneLivesAnimation: Bool

    init(
        screenState: BasicScreenState = .loading,
        gameData: GameData = GameData.companion.initial(),
        gameId: Int = -1,
        gridRowsState: [Int: GameGridRowState] = [:],
        currentSelection: [GameDistributionCoordinates] = [],
        lives: Int = -1,
        selectFailed: Bool = false,
        livesEarnedOnDone: Int = 0,
        nextGameId: Int = -1,
        notSolvedRowsOnLost: [Int: Int] = [:],
        showOnDoneSignAnimation: Bool = false,
        showOnDoneLivesAnimation: Bool = false
    ) {
        self.screenState = screenState
        self.gameData = gameData
        self.gameId = gameId
        self.gridRowsState = gridRowsState
        self.currentSelection = currentSelection
        self.lives = lives
        self.selectFailed = selectFailed
        self.livesEarnedOnDone = livesEarnedOnDone
        self.nextGameId = nextGameId
        self.notSolvedRowsOnLost = notSolvedRowsOnLost
        self.showOnDoneSignAnimation = showOnDoneSignAnimation
        self.showOnDoneLivesAnimation = showOnDoneLivesAnimation
    }

    var canSelectWords: Bool {
        gameData.gameState == GameState.playing && currentSelection.count != GameContractsKt.GAME_WORD_AMOUNT
    }

    var canDeselectWords: Bool {
        gameData.gameState == GameState.playing && currentSelection.isEmpty
    }

    var canSendRow: Bool {
        gameData.gameState == GameState.playing && !selectFailed && currentSelection.count == GameContractsKt.GAME_WORD_AMOUNT
    }

    var showButtonsOnDone: Bool {
        let state = gameData.gameState

        if (state != .win && state != .lost) || nextGameId == -1 {
            return false
        }

        return state == .win ? !showOnDoneSignAnimation && !showOnDoneLivesAnimation : !showOnDoneSignAnimation
    }
}
